compiler/qsc_frontend/src/error.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#[cfg(test)]
+mod tests;
+
+use crate::compile::{Source, SourceMap};
+use miette::{Diagnostic, MietteError, MietteSpanContents, SourceCode, SourceSpan, SpanContents};
+use std::{
+    error::Error,
+    fmt::{self, Debug, Display, Formatter},
+};
+
+#[derive(Clone, Debug)]
+pub struct WithSource<E>
```

Comment:
`WithSource` is pre-existing, I just reworked it quite a bit. Now, the only way to instantiate it is with `from_map` which takes a `SourceMap`.

I also removed the optional stack trace field and moved it to a new struct definition (`WithStack` above).